guides-markdown/swaps.md

```diff
@@ -0,0 +1,650 @@
+Bob thinks Alice is going to flake out, he can refund his Bitcoin from Mailbox “A” long before Alice can
+refund her Bitcoin Cash from Mailbox “B”. 
+
+Think about the counter-example. If Mailbox “B” had the shorter refund time, Alice could wait until that time
+expires, refund herself the Bitcoin Cash from Mailbox “B” AND THEN ALSO enter the secret code into Bitcoin
+Mailbox “A” and take the Bitcoin that Bob sent. Alice would have all the money and Bob would be broke!
+
+<a name="scripts"></a>
+## Scripts
+
+Now it's time to turn these "two-lock mailboxes" into actual Bitcoin transactions. For this guide we
+will describe a very simple method that sacrifices some privacy and potentially some security. See
+[next steps](#next-steps) for more details on this.
+This will be the redeem script in the output of our Hash Time-Locked Contract:
+
+```bash
+OP_IF
+  OP_SHA256
+  <hash of secret>
+  OP_EQUALVERIFY
+  <pubKey of swap>
+  OP_CHECKSIG
+OP_ELSE
+  <relative locktime>
+  OP_CHECKSEQUENCEVERIFY
+  OP_DROP
+  <pubKey of refund>
+  OP_CHECKSIG
+OP_ENDIF
+```
+
+The Hash Lock is the first section of the script, executed when `OP_IF` reads a `true` value from the stack.
+It expects a secret value and an ECDSA signature. It hashes the secret and checks that it matches a given hash,
+then it checks the signature against the given public key. To redeem the HTLC this way, we will use the following
+script in the input of a transaction:
+
+```bash
+<signature of swap>
+<secret value>
+<true>
+```
+
+Keep in mind that the stack is first-in / last-out, so if this input script looks "upside-down" that's because the
+last element will end up being on the top of the stack when the redeem script is run.
+
+The Time Lock is executed when `OP_IF` gets a `false` value. First it executes a
+[Check Sequence Verify](https://github.com/bitcoinbook/bitcoinbook/blob/develop/ch07.asciidoc#relative-timelocks)
+routine, which fails if the given relative locktime has not yet passed. Then it simply checks a signature
+against a given public key like usual. To redeem the transaction with this path, we will use this input script:
+
+```bash
+<signature of refund>
+<false>
+```
+
+## bcoin and bcash
+
+The `bcoin` and `bcash` libraries are so similar that we can use the exact same commands
+to create HTLCs on both chains, and build a working application with wallets. A finished package
+can be found at [https://github.com/pinheadmz/swap-test](https://github.com/pinheadmz/swap-test). First, we'll
+review the `Swap` class in `lib/swap.js` that is able to build scripts, addresses, and transactions
+on any compatible library passed to it. With almost no conditionals, we can make this library completely
+chain-agnostic.
+
+```javascript
+// lib/swap.js
+
+class Swap {
+  constructor(lib, network){
+    // Require the library by name and set the network for every module
+    this.libName = lib;
```

Comment:
This is a good point. In `app/run-swap.js` I check all the input params, but left that stuff out of the guide just for brevity and clarity. I could add checks in `lib/swap.js` but actually, I kinda don't need to because these lines will fail on their own if bad input is passed:

```
    this.lib = require(lib);
    this.lib.Network.set(network);
```